### PR TITLE
Gitstatusd support take 2

### LINKED ISSUE #1378: Integration with gitstatusd
This issue is to gauge interest in integrating Bash-it with [gitstatus](https://github.com/romkatv/gitstatus/blob/master/README.md#using-from-bash).

Bash themes normally call `git` commands to obtain information about the current Git repository. This can be slow when working with large repositories and/or using underpowered devices such as Raspberry Pi. [gitstatus](https://github.com/romkatv/gitstatus/blob/master/README.md#using-from-bash) provides an alternative way to extract the same information 10 times faster.

Is this something you would find useful in Bash-it?

### PR DESCRIPTION
**Pull request by NoahGorny**
Resolves #1378
Second take on #1429 attempt which I took inspiration (and commits) from, thanks @markusdd!

TODO:
- [x] Fix unwanted usage after disabling the plugin (need to `unset` the `USE_GITSTATUSD` flag)
- [x] Add tests? (tested the plugin myself, pretty sure thats enough)

### CODE REVIEW

**Comment by markusdd** [comment]
@NoahGorny Thanks so very much for picking this up!

Actually I've been using my version successfully for quite some time, and I'm very happy to see it embedded properly now using the plugin mechanics, which I was not very familiar with.

Really would like to see this merged to mainline.

One question though: Can we make the users life easier be providing a cross-link or little helper routine to install gitstatusd from @romkatv repository https://github.com/romkatv/gitstatus ?

**Comment by NoahGorny** [comment]
> @NoahGorny Thanks so very much for picking this up!
> 
> Actually I've been using my version successfully for quite some time, and I'm very happy to see it embedded properly now using the plugin mechanics, which I was not very familiar with.
> 
> Really would like to see this merged to mainline.
> 
> One question though: Can we make the users life easier be providing a cross-link or little helper routine to install gitstatusd from @romkatv repository https://github.com/romkatv/gitstatus ?

Thank you for your kind words and support :smile:
I added a ref in the `README.md` and in the plugin description to the repo. I assume bash-it users can clone and manage stuff like this themselves, but I am open to ideas :)

File: plugins/available/gitstatusd.plugin.bash [hunk 1]
```diff
@@ -0,0 +1,22 @@
+cite about-plugin
+about-plugin 'speeds up your life by using gitstatusd for git status calculations. install from https://github.com/romkatv/gitstatus'
+
+
+function gitstatusd_on_disable() {
+  about 'Destructor of gitstatud plugin'
+  group 'gitstatud'
```
Comment (romkatv) [inline plugins/available/gitstatusd.plugin.bash:7]: Typo in the name.

File: plugins/available/gitstatusd.plugin.bash [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+  unset SCM_GIT_USE_GITSTATUSD
```
Comment (romkatv) [inline plugins/available/gitstatusd.plugin.bash:9]: Perhaps also call `gitstatus_stop`?

File: plugins/available/gitstatusd.plugin.bash [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+}
+
+[[ "$SCM_CHECK" == "true" ]] || return # No scm-check
+
+[[ $- == *i* ]] || return  # non-interactive shell
+
+[[ ! -z "${SCM_GIT_GITSTATUSD_LOC}" ]] || SCM_GIT_GITSTATUSD_LOC="$HOME/gitstatus/gitstatus.plugin.sh"
```
Comment (romkatv) [inline plugins/available/gitstatusd.plugin.bash:16]: This is usually done like this:

```sh
: "${SCM_GIT_GITSTATUSD_LOC:=$HOME/gitstatus/gitstatus.plugin.sh}"
```

You might also want to rename this parameter as it doesn't actually point to gitstatusd.

File: plugins/available/gitstatusd.plugin.bash [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+if [[ -f "${SCM_GIT_GITSTATUSD_LOC}" ]]; then
+  source "${SCM_GIT_GITSTATUSD_LOC}"
+  # Start the actual gitstatud binary
```
Comment (romkatv) [inline plugins/available/gitstatusd.plugin.bash:19]: Typo in the name.

File: template/bash_profile.template.bash [hunk 5]
```diff
@@ -32,6 +32,10 @@ export TODO="t"
 
 # Set this to false to turn off version control status checking within the prompt for all themes
 export SCM_CHECK=true
+# Set to actual location of gitstatusd if installed
```
Comment (romkatv) [inline template/bash_profile.template.bash:35]: Same here: It's not pointing to gitstatusd.

File: themes/base.theme.bash [hunk 6]
```diff
@@ -191,6 +194,12 @@ function git_prompt_minimal_info {
 }
 
 function git_prompt_vars {
+  if ${SCM_GIT_USE_GITSTATUSD} && _command_exists gitstatus_query && gitstatus_query && [[ "${VCS_STATUS_RESULT}" == "ok-sync" ]]; then # use faster gitstatusd
```
Comment (romkatv) [inline themes/base.theme.bash:197]: `${SCM_GIT_USE_GITSTATUSD}` is true if unset. Is this what you want?

File: themes/githelpers.theme.bash [hunk 7]
```diff
@@ -101,29 +117,51 @@ function _git-status-counts {
...
-  [[ "$(_git-branch)" == "$(_git-upstream-branch)" ]] && same_branch_name=true
-  if ([[ "${SCM_GIT_SHOW_REMOTE_INFO}" = "auto" ]] && [[ "$(_git-num-remotes)" -ge 2 ]]) ||
-      [[ "${SCM_GIT_SHOW_REMOTE_INFO}" = "true" ]]; then
-    if [[ "${same_branch_name}" != "true" ]]; then
-      remote_info="\$(_git-upstream)"
-    else
-      remote_info="$(_git-upstream-remote)"
+
+  if [[ "${SCM_GIT_GITSTATUSD_RAN}" == "true" ]]; then # prompt handling only, reimplement because patching the routine below gets ugly
+    [[ "${VCS_STATUS_REMOTE_NAME}" == "" ]] && return || true
```
Comment (romkatv) [inline themes/githelpers.theme.bash:122]: This `|| true` has no effect. The same in several other places.

File: themes/githelpers.theme.bash [hunk 8]
```diff
@@ -101,29 +117,51 @@ function _git-status-counts {
...
+    [[ "${VCS_STATUS_LOCAL_BRANCH}" == "${VCS_STATUS_REMOTE_BRANCH}" ]] && local same_branch_name=true || true
+    local same_branch_name=
+    [[ "${VCS_STATUS_LOCAL_BRANCH}" == "${VCS_STATUS_REMOTE_BRANCH}" ]] && same_branch_name=true
+    if [[ "${SCM_GIT_SHOW_REMOTE_INFO}" = "true" ]] || [[ "${SCM_GIT_SHOW_REMOTE_INFO}" = "auto" ]]; then # no multiple remote support in gitstatusd
```
Comment (romkatv) [inline themes/githelpers.theme.bash:126]: You can replace `]] || [[` with just `||`.

Comment (NoahGorny) [reply themes/githelpers.theme.bash]: done

File: themes/githelpers.theme.bash (hunk 7 above)
Comment (NoahGorny) [reply themes/githelpers.theme.bash]: done

File: themes/base.theme.bash (hunk 6 above)
Comment (NoahGorny) [reply themes/base.theme.bash:196]: It will be set by `base.theme.bash`

File: template/bash_profile.template.bash (hunk 5 above)
Comment (NoahGorny) [reply template/bash_profile.template.bash]: done

File: plugins/available/gitstatusd.plugin.bash (hunk 4 above)
Comment (NoahGorny) [reply plugins/available/gitstatusd.plugin.bash]: done

File: plugins/available/gitstatusd.plugin.bash (hunk 3 above)
Comment (NoahGorny) [reply plugins/available/gitstatusd.plugin.bash]: done and renamed

File: plugins/available/gitstatusd.plugin.bash (hunk 2 above)
Comment (NoahGorny) [reply plugins/available/gitstatusd.plugin.bash:9]: done

File: plugins/available/gitstatusd.plugin.bash (hunk 1 above)
Comment (NoahGorny) [reply plugins/available/gitstatusd.plugin.bash]: done

**Comment by markusdd** [comment]
I actually noticed @romkatv did some nice improvements in the meantime as well, so gitstatusd will pull in the correct binary once the plugin is loaded.

That should be convenient enough.

File: README.md [hunk 9]
```diff
@@ -395,6 +395,13 @@ $ git config --global --add bash-it.hide-status 1
 
 Setting this flag globally has the same effect as `SCM_CHECK=true`, but only for Git repos.
 
+### Speed up git status calculations
+
+As an alternative to ignoring repo status entierly, you can try out the `gitstatusd` plugin.
+This plugin speeds up all `git status` calculations by up to 10x times!
+
+**NOTE**: You will need to clone `gitstatusd` repo from [here](https://github.com/romkatv/gitstatus).
```
Comment (romkatv) [inline README.md:403]: The repo is called `gitstatus` (no `d` at the end).

**Comment by NoahGorny** [comment]
Btw guys for some reason the CI fails on mac, it seems like the order of the results of `search: git` is different on mac from ubuntu. On my machine `gitstatusd` plugin comes first, in mac `git-subrepo` comes first
If you have any idea why I would love to get some help @markusdd @nwinkler

**Comment by markusdd** [comment]
mmh I do have a mac mini at hand but I run zsh on that one.

let me check the logs

**Comment by markusdd** [comment]
Mmh indeed:
image

Question is: what impact does this order have?

**Comment by NoahGorny** [comment]
No impact, just failing the tests
I assume I can rewrite this test and fix this

I also consider adding tests to check that the plugin works correctly, but that will be hard

**Comment by markusdd** [comment]
I assume its the encoding on the machines, apparently dashes have a different order, but nothing new there, on macs everything has to be different it seems...

**Comment by NoahGorny** [comment]
I see, I will rename it in my spare time today

**Comment by markusdd** [comment]
Looks all good to me now, also thanks @romkatv for reviewing again.

**Comment by NoahGorny** [comment]
@romkatv take a final look now, I renamed the plugin.
Also, thank you @markusdd  and @romkatv for your valuable feedback!

**Comment by markusdd** [comment]
> I think it would be better to replace this parameter with `SCM_GIT_GITSTATUS_DIR`:
> 
> ```shell
> : ${SCM_GIT_GITSTATUS_DIR:="$HOME/gitstatus"}
> ```
> 
> (Also note that I've removed redundant self-assignment.)
> 
> There is more than one file in the gitstatus directory different files depend on each other. `SCM_GIT_GITSTATUS_DIR` makes it clear that you aren't supposed to pluck an individual file.

Fair point.

**Comment by NoahGorny** [comment]
> I think it would be better to replace this parameter with `SCM_GIT_GITSTATUS_DIR`:
> 
> ```shell
> : ${SCM_GIT_GITSTATUS_DIR:="$HOME/gitstatus"}
> ```
> 
> (Also note that I've removed redundant self-assignment.)
> 
> There is more than one file in the gitstatus directory different files depend on each other. `SCM_GIT_GITSTATUS_DIR` makes it clear that you aren't supposed to pluck an individual file.

done

File: README.md [hunk 10]
```diff
@@ -395,6 +395,13 @@ $ git config --global --add bash-it.hide-status 1
 
 Setting this flag globally has the same effect as `SCM_CHECK=true`, but only for Git repos.
 
+### Speed up git status calculations
+
+As an alternative to ignoring repo status entierly, you can try out the `gitstatus` plugin.
```
Comment (nwinkler) [inline README.md:400]: Typo in `entirely`.

File: lib/helpers.bash [hunk 11]
```diff
@@ -326,6 +337,7 @@ _disable-plugin ()
     _group 'lib'
 
     _disable-thing "plugins" "plugin" $1
+    _on-disable-callback $1
```
Comment (nwinkler) [inline lib/helpers.bash:340]: This is great - love this idea! I think it makes sense to provide a note about this in the `DEVELOPMENT.md` file.

File: plugins/available/gitstatus.plugin.bash [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+
+function gitstatus_on_disable() {
+  about 'Destructor of gitstatus plugin'
+  group 'gitstatus'
+
+  unset SCM_GIT_USE_GITSTATUS
+  _command_exists gitstatus_stop && gitstatus_stop
+}
+
+[[ "$SCM_CHECK" == "true" ]] || return # No scm-check
```
Comment (nwinkler) [inline plugins/available/gitstatus.plugin.bash:13]: Minor comment - can you please move comments like here to the line above? Same for the command below...

Comment (NoahGorny) [reply plugins/available/gitstatus.plugin.bash]: done

File: lib/helpers.bash (hunk 11 above)
Comment (NoahGorny) [reply lib/helpers.bash:340]: done

File: README.md (hunk 10 above)
Comment (NoahGorny) [reply README.md]: done

**Comment by nwinkler** [comment]
Thanks for the great work, @NoahGorny!

Thanks for the thorough reviews, @markusdd and @romkatv!

Merging this now.

**Comment by NoahGorny** [comment]
Thanks a lot @nwinkler !
I would be grateful if you could take a look on the other PRs that I have opened, because I think they are pretty nice :smile:

**Comment by nwinkler** [comment]
Going through them one by one - thanks! I will probably need a couple of days to go through all of them, I can only use a limited amount of time each day for this...

Keep them coming, we will work through them, promised!

**Comment by markusdd** [comment]
Is my original PR closed with this or do we need to do that manually?

@NoahGorny I believe you merged my initial changes to your branch and built on that right?

**Comment by nwinkler** [comment]
I closed #1429 yesterday after merging this PR.

**Comment by NoahGorny** [comment]
> Is my original PR closed with this or do we need to do that manually?
> 
> @NoahGorny I believe you merged my initial changes to your branch and built on that right?

I indeed took your original commits and built on it :smile:

**Comment by markusdd** [comment]
Thanks Guys!

**Pull request merged**